File: packages/ffi/android/testapp/src/main/kotlin/dev/wvb/testapp/TestRunner.kt
package dev.wvb.testapp

import android.content.Context
import dev.wvb.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File

data class TestResult(
    val name: String,
    val passed: Boolean,
    val error: String? = null,
)

class TestRunner(private val context: Context) {
    private val indexHtml = "<!DOCTYPE html>".toByteArray()
    private val indexJs = "console.log('hello')".toByteArray()

    suspend fun run(): List<TestResult> = withContext(Dispatchers.IO) {
        val results = mutableListOf<TestResult>()

        fun test(name: String, block: () -> Unit) {
            try {
                block()
                results.add(TestResult(name, true))
            } catch (e: Throwable) {
                results.add(TestResult(name, false, e.message ?: e.javaClass.simpleName))
            }
        }

        suspend fun testSuspend(name: String, block: suspend () -> Unit) {
            try {
                block()
                results.add(TestResult(name, true))
            } catch (e: Throwable) {
                results.add(TestResult(name, false, e.message ?: e.javaClass.simpleName))
            }
        }

        // ── HttpResponse ─────────────────────────────────────────────────
        test("HttpResponse: init") {
            val response = HttpResponse(
                status = 200u,
                headers = mapOf("Content-Type" to "text/html"),
                body = byteArrayOf()
            )
            check(response.status == 200.toUShort()) { "status mismatch" }
            check(response.headers["Content-Type"] == "text/html") { "header mismatch" }
        }

        // ── BundleBuilder ────────────────────────────────────────────────
        test("BundleBuilder: build") {
            val builder = BundleBuilder(null)
            check(builder.version() == Version.V1)
            check(builder.entryPaths().isEmpty())
            check(!builder.insertEntry("/index.js", indexJs, null, null)) { "first insert should return false" }
            check(!builder.insertEntry("/index.html", indexHtml, null, null))
            check(builder.insertEntry("/index.html", indexHtml, null, null)) { "duplicate insert should return true" }
            check(builder.entryPaths().size == 2)
            check(builder.containsEntry("/index.js"))
            check(builder.containsEntry("/index.html"))
            check(!builder.containsEntry("/not_exists"))
            builder.build(null)
        }

        test("BundleBuilder: with options") {
            val builder = BundleBuilder(Version.V1)
            builder.insertEntry("/index.js", indexJs, null, null)
            builder.build(
                BuildOptions(
                    header = BuildHeaderOptions(checksum = ChecksumWriteOptions(seed = 1u)),
                    index = BuildIndexOptions(checksum = ChecksumWriteOptions(seed = 2u)),
                    dataChecksum = ChecksumWriteOptions(seed = 3u),
                )
            )
        }

        test("BundleBuilder: removeEntry") {
            val builder = BundleBuilder(null)
            builder.insertEntry("/index.js", indexJs, null, null)
            check(builder.removeEntry("/index.js"))
            check(!builder.containsEntry("/index.js"))
            check(!builder.removeEntry("/index.js"))
        }

        // ── Bundle ───────────────────────────────────────────────────────
        test("Bundle: version") {
            val bundle = BundleBuilder(Version.V1).build(null)
            check(bundle.descriptor().header().version() == Version.V1)
        }

        test("Bundle: getData") {
            val builder = BundleBuilder(null)
            builder.insertEntry("/index.js", indexJs, null, null)
            builder.insertEntry("/index.html", indexHtml, null, null)
            val bundle = builder.build(null)
            check(bundle.getData("/index.js").contentEquals(indexJs))
            check(bundle.getData("/index.html").contentEquals(indexHtml))
            check(bundle.getData("/not_exists") == null)
        }

        test("Bundle: index entries") {
            val builder = BundleBuilder(null)
            builder.insertEntry("/index.js", indexJs, null, null)
            builder.insertEntry("/index.html", indexHtml, null, null)
            val bundle = builder.build(null)
            val index = bundle.descriptor().index()
            check(index.entries().size == 2)
            check(index.containsPath("/index.js"))
            check(index.containsPath("/index.html"))
            check(!index.containsPath("/not_exists"))
            check(index.getEntry("/index.js") != null)
            check(index.getEntry("/not_exists") == null)
        }

        test("Bundle: index entry metadata") {
            val builder = BundleBuilder(null)
            builder.insertEntry("/index.js", indexJs, "text/javascript", mapOf("x-custom" to "value"))
            val bundle = builder.build(null)
            val entry = bundle.descriptor().index().getEntry("/index.js")
            checkNotNull(entry)
            check(entry.contentType == "text/javascript") { "contentType=${entry.contentType}" }
            check(entry.contentLength == indexJs.size.toULong())
            check(entry.headers["x-custom"] == "value")
        }

        test("Bundle: descriptor indexEntries") {
            val builder = BundleBuilder(null)
            builder.insertEntry("/index.js", indexJs, null, null)
            val bundle = builder.build(null)
            val descriptor = bundle.descriptor()
            check(descriptor.indexEntries().size == 1)
            check(descriptor.containsPath("/index.js"))
            check(!descriptor.containsPath("/not_exists"))
            check(descriptor.getIndexEntry("/index.js") != null)
        }

        test("Bundle: getDataChecksum") {
            val builder = BundleBuilder(null)
            builder.insertEntry("/index.js", indexJs, null, null)
            val bundle = builder.build(null)
            check(bundle.getDataChecksum("/index.js") != null)
            check(bundle.getDataChecksum("/not_exists") == null)
        }

        // ── Read/Write ───────────────────────────────────────────────────
        test("Read/Write: bytes roundtrip") {
            val builder = BundleBuilder(null)
            builder.insertEntry("/index.js", indexJs, null, null)
            builder.insertEntry("/index.html", indexHtml, null, null)
            val bundle = builder.build(null)

            val bytes = writeBundleToBytes(bundle)
            check(bytes.isNotEmpty())

            val loaded = readBundleFromBytes(bytes)
            check(loaded.descriptor().header().version() == Version.V1)
            check(loaded.getData("/index.js").contentEquals(indexJs))
            check(loaded.getData("/index.html").contentEquals(indexHtml))
        }

        testSuspend("Read/Write: file roundtrip") {
            val tmpFile = File(context.filesDir, "test_bundle.wvb")
            try {
                val builder = BundleBuilder(null)
                builder.insertEntry("/index.js", indexJs, null, null)
                builder.insertEntry("/index.html", indexHtml, null, null)
                val bundle = builder.build(null)

                writeBundle(bundle, tmpFile.absolutePath)
                check(tmpFile.exists() && tmpFile.length() > 0)

                val loaded = readBundle(tmpFile.absolutePath)
                check(loaded.descriptor().header().version() == Version.V1)
                val index = loaded.descriptor().index()
                check(index.entries().size == 2)
                check(index.containsPath("/index.js"))
                check(loaded.getData("/index.js").contentEquals(indexJs))
            } finally {
                tmpFile.delete()
            }
        }

        // ── Protocol ─────────────────────────────────────────────────────
        testSuspend("BundleProtocolHandler: 200 index.html") {
            withSource { source ->
                val handler = BundleProtocolHandler(source)
                val response = handler.handle(HttpMethod.GET, "https://app.wvb/index.html", null)
                check(response.status == 200.toUShort()) { "status=${response.status}" }
                check(response.body.isNotEmpty()) { "body should not be empty" }
                check(response.headers.containsKey("content-type")) { "content-type header missing" }
            }
        }

        testSuspend("BundleProtocolHandler: 200 root redirect") {
            withSource { source ->
                val handler = BundleProtocolHandler(source)
                val response = handler.handle(HttpMethod.GET, "https://app.wvb/", null)
                check(response.status == 200.toUShort()) { "status=${response.status}" }
            }
        }

        testSuspend("BundleProtocolHandler: 404 not found") {
            withSource { source ->
                val handler = BundleProtocolHandler(source)
                val response = handler.handle(HttpMethod.GET, "https://app.wvb/not_found.html", null)
                check(response.status == 404.toUShort()) { "status=${response.status}" }
            }
        }

        testSuspend("BundleProtocolHandler: HEAD 200") {
            withSource { source ->
                val handler = BundleProtocolHandler(source)
                val response = handler.handle(HttpMethod.HEAD, "https://app.wvb/index.html", null)
                check(response.status == 200.toUShort()) { "status=${response.status}" }
                check(response.body.isEmpty()) { "HEAD response body must be empty" }
            }
        }

        testSuspend("BundleProtocolHandler: exact path resolver does not rewrite to index.html") {
            withSource { source ->
                val options = BundleProtocolOptions(pathResolver = PathResolver.EXACT)
                val handler = BundleProtocolHandler(source, options)
                check(handler.handle(HttpMethod.GET, "https://app.wvb/index.html", null).status == 200.toUShort())
                val response = handler.handle(HttpMethod.GET, "https://app.wvb/", null)
                check(response.status == 404.toUShort()) { "status=${response.status}" }
            }
        }

        testSuspend("BundleProtocolHandler: allowWvbSuffixOnly rejects other hosts") {
            withSource { source ->
                val options = BundleProtocolOptions(
                    bundleResolver = BundleResolver.Hostname(
                        segment = HostnameSegment.First,
                        allowWvbSuffixOnly = true,
                    ),
                )
                val handler = BundleProtocolHandler(source, options)
                check(handler.handle(HttpMethod.GET, "https://app.wvb/index.html", null).status == 200.toUShort())
                // Recorded outside the catch, which would swallow an `error(...)` raised inside it.
                var rejected = false
                try {
                    handler.handle(HttpMethod.GET, "https://app.example.com/index.html", null)
                } catch (e: Exception) {
                    // expected: the bundle name is not resolved, so no bundle is found
                    rejected = true
                }
                check(rejected) { "expected an error for a host without the .wvb suffix" }
            }
        }

        test("ProxyProtocolHandler: init") {
            ProxyProtocolHandler(mapOf("myapp" to "http://localhost:9999"))
        }

        testSuspend("BundleProtocolHandler: a request body is accepted") {
            withSource { source ->
                val handler = BundleProtocolHandler(source)
                // The bundle protocol serves GET/HEAD only, but the body still travels over the FFI.
                val response = handler.handle(
                    HttpMethod.POST,
                    "https://app.wvb/index.html",
                    null,
                    """{"hello":"world"}""".toByteArray(),
                )
                check(response.status == 405.toUShort()) { "status=${response.status}" }
            }
        }

        testSuspend("ProxyProtocolHandler: unknown host error") {
            val handler = ProxyProtocolHandler(mapOf("known" to "http://localhost:9999"))
            var rejected = false
            try {
                handler.handle(HttpMethod.GET, "https://unknown.wvb/index.html", null)
            } catch (e: Exception) {
                // expected: no proxy target for "unknown.wvb"
                rejected = true
            }
            check(rejected) { "expected an error for an unknown host" }
        }

        testSuspend("ProxyProtocolHandler: custom resolver receives the uri") {
            var seen: String? = null
            val resolver = object : ProxyResolver {
                override suspend fun resolve(uri: String): String? {
                    seen = uri
                    return null // do not proxy
                }
            }
            val handler = ProxyProtocolHandler.custom(resolver)
            var rejected = false
            try {
                handler.handle(HttpMethod.GET, "https://app.wvb/index.html", null)
            } catch (e: Exception) {
                // expected: the target is unresolved
                rejected = true
            }
            check(rejected) { "expected an error when the resolver returns null" }
            check(seen == "https://app.wvb/index.html") { "resolver saw uri=$seen" }
        }

        results
    }

    private fun setupFixtures(): File {
        val remoteDir = File(context.filesDir, "fixture_remote")
        if (!File(remoteDir, "manifest.json").exists()) {
            remoteDir.mkdirs()
            context.assets.open("fixtures/remote/manifest.json").use { input ->
                File(remoteDir, "manifest.json").outputStream().use { input.copyTo(it) }
            }
            val appDir = File(remoteDir, "app")
            appDir.mkdirs()
            context.assets.open("fixtures/remote/app/1.0.0.wvb").use { input ->
                File(appDir, "1.0.0.wvb").outputStream().use { input.copyTo(it) }
            }
        }
        return remoteDir
    }

    private suspend fun withSource(block: suspend (Source) -> Unit) {
        val remoteDir = setupFixtures()
        val source = Source(
            SourceConfig(
                builtinDir = context.cacheDir.absolutePath,
                remoteDir = remoteDir.absolutePath,
                builtinManifestFilepath = null,
                remoteManifestFilepath = null,
            )
        )
        block(source)
    }
}

File: packages/ffi/apple/ios/TestApp/TestRunner.swift
import Foundation
import WebViewBundleLibrary

struct TestResult: Identifiable {
    let id = UUID()
    let name: String
    let passed: Bool
    let error: String?
}

@MainActor
final class TestRunner: ObservableObject {
    @Published var results: [TestResult] = []
    @Published var isRunning = false

    private let indexHtml = Data("<!DOCTYPE html>".utf8)
    private let indexJs = Data("console.log('hello')".utf8)

    func run() async {
        isRunning = true
        results = []

        var newResults: [TestResult] = []

        func test(_ name: String, _ block: () throws -> Void) {
            do {
                try block()
                newResults.append(TestResult(name: name, passed: true, error: nil))
            } catch {
                newResults.append(TestResult(name: name, passed: false, error: error.localizedDescription))
            }
        }

        func testAsync(_ name: String, _ block: () async throws -> Void) async {
            do {
                try await block()
                newResults.append(TestResult(name: name, passed: true, error: nil))
            } catch {
                newResults.append(TestResult(name: name, passed: false, error: error.localizedDescription))
            }
        }

        // ── HttpResponse ─────────────────────────────────────────────────
        test("HttpResponse: init") {
            let response = HttpResponse(status: 200, headers: ["Content-Type": "text/html"], body: Data())
            guard response.status == 200 else { throw Fail("status mismatch") }
            guard response.headers["Content-Type"] == "text/html" else { throw Fail("header mismatch") }
        }

        // ── BundleBuilder ────────────────────────────────────────────────
        test("BundleBuilder: build") {
            let builder = BundleBuilder(version: nil)
            guard builder.version() == .v1 else { throw Fail("version mismatch") }
            guard builder.entryPaths().isEmpty else { throw Fail("expected empty entryPaths") }
            let ins1 = try builder.insertEntry(path: "/index.js", data: indexJs, contentType: nil, headers: nil)
            guard ins1 == false else { throw Fail("first insert should return false") }
            let ins2 = try builder.insertEntry(path: "/index.html", data: indexHtml, contentType: nil, headers: nil)
            guard ins2 == false else { throw Fail("second insert should return false") }
            let ins3 = try builder.insertEntry(path: "/index.html", data: indexHtml, contentType: nil, headers: nil)
            guard ins3 == true else { throw Fail("duplicate insert should return true") }
            guard builder.entryPaths().count == 2 else { throw Fail("expected 2 entries") }
            guard builder.containsEntry(path: "/index.js") else { throw Fail("missing /index.js") }
            guard builder.containsEntry(path: "/index.html") else { throw Fail("missing /index.html") }
            guard !builder.containsEntry(path: "/not_exists") else { throw Fail("/not_exists should not exist") }
            _ = try builder.build(options: nil)
        }

        test("BundleBuilder: with options") {
            let builder = BundleBuilder(version: .v1)
            _ = try builder.insertEntry(path: "/index.js", data: indexJs, contentType: nil, headers: nil)
            _ = try builder.build(options: BuildOptions(
                header: BuildHeaderOptions(checksum: ChecksumWriteOptions(seed: 1)),
                index: BuildIndexOptions(checksum: ChecksumWriteOptions(seed: 2)),
                dataChecksum: ChecksumWriteOptions(seed: 3)
            ))
        }

        test("BundleBuilder: removeEntry") {
            let builder = BundleBuilder(version: nil)
            _ = try builder.insertEntry(path: "/index.js", data: indexJs, contentType: nil, headers: nil)
            guard builder.removeEntry(path: "/index.js") else { throw Fail("removeEntry should return true") }
            guard !builder.containsEntry(path: "/index.js") else { throw Fail("entry should be removed") }
            guard !builder.removeEntry(path: "/index.js") else { throw Fail("second remove should return false") }
        }

        // ── Bundle ───────────────────────────────────────────────────────
        test("Bundle: version") {
            let bundle = try BundleBuilder(version: .v1).build(options: nil)
            guard bundle.descriptor().header().version() == .v1 else { throw Fail("version mismatch") }
        }

        test("Bundle: getData") {
            let builder = BundleBuilder(version: nil)
            _ = try builder.insertEntry(path: "/index.js", data: indexJs, contentType: nil, headers: nil)
            _ = try builder.insertEntry(path: "/index.html", data: indexHtml, contentType: nil, headers: nil)
            let bundle = try builder.build(options: nil)
            guard try bundle.getData(path: "/index.js") == indexJs else { throw Fail("index.js data mismatch") }
            guard try bundle.getData(path: "/index.html") == indexHtml else { throw Fail("index.html data mismatch") }
            guard try bundle.getData(path: "/not_exists") == nil else { throw Fail("/not_exists should return nil") }
        }

        test("Bundle: index entries") {
            let builder = BundleBuilder(version: nil)
            _ = try builder.insertEntry(path: "/index.js", data: indexJs, contentType: nil, headers: nil)
            _ = try builder.insertEntry(path: "/index.html", data: indexHtml, contentType: nil, headers: nil)
            let bundle = try builder.build(options: nil)
            let index = bundle.descriptor().index()
            guard index.entries().count == 2 else { throw Fail("expected 2 index entries") }
            guard index.containsPath(path: "/index.js") else { throw Fail("missing /index.js") }
            guard index.containsPath(path: "/index.html") else { throw Fail("missing /index.html") }
            guard !index.containsPath(path: "/not_exists") else { throw Fail("/not_exists should not exist") }
            guard index.getEntry(path: "/index.js") != nil else { throw Fail("getEntry returned nil") }
            guard index.getEntry(path: "/not_exists") == nil else { throw Fail("getEntry should return nil") }
        }

        test("Bundle: index entry metadata") {
            let builder = BundleBuilder(version: nil)
            _ = try builder.insertEntry(path: "/index.js", data: indexJs, contentType: "text/javascript", headers: ["x-custom": "value"])
            let bundle = try builder.build(options: nil)
            guard let entry = bundle.descriptor().index().getEntry(path: "/index.js") else { throw Fail("entry not found") }
            guard entry.contentType == "text/javascript" else { throw Fail("contentType=\(entry.contentType)") }
            guard entry.contentLength == UInt64(indexJs.count) else { throw Fail("contentLength mismatch") }
            guard entry.headers["x-custom"] == "value" else { throw Fail("header mismatch") }
        }

        test("Bundle: descriptor indexEntries") {
            let builder = BundleBuilder(version: nil)
            _ = try builder.insertEntry(path: "/index.js", data: indexJs, contentType: nil, headers: nil)
            let bundle = try builder.build(options: nil)
            let descriptor = bundle.descriptor()
            guard descriptor.indexEntries().count == 1 else { throw Fail("expected 1 entry") }
            guard descriptor.containsPath(path: "/index.js") else { throw Fail("missing /index.js") }
            guard !descriptor.containsPath(path: "/not_exists") else { throw Fail("/not_exists should not exist") }
            guard descriptor.getIndexEntry(path: "/index.js") != nil else { throw Fail("getIndexEntry returned nil") }
        }

        test("Bundle: getDataChecksum") {
            let builder = BundleBuilder(version: nil)
            _ = try builder.insertEntry(path: "/index.js", data: indexJs, contentType: nil, headers: nil)
            let bundle = try builder.build(options: nil)
            guard try bundle.getDataChecksum(path: "/index.js") != nil else { throw Fail("checksum should not be nil") }
            guard try bundle.getDataChecksum(path: "/not_exists") == nil else { throw Fail("checksum should be nil") }
        }

        // ── Read/Write ───────────────────────────────────────────────────
        test("Read/Write: bytes roundtrip") {
            let builder = BundleBuilder(version: nil)
            _ = try builder.insertEntry(path: "/index.js", data: indexJs, contentType: nil, headers: nil)
            _ = try builder.insertEntry(path: "/index.html", data: indexHtml, contentType: nil, headers: nil)
            let bundle = try builder.build(options: nil)
            let bytes = try writeBundleToBytes(bundle: bundle)
            guard !bytes.isEmpty else { throw Fail("bytes should not be empty") }
            let loaded = try readBundleFromBytes(data: bytes)
            guard loaded.descriptor().header().version() == .v1 else { throw Fail("version mismatch") }
            guard try loaded.getData(path: "/index.js") == indexJs else { throw Fail("index.js data mismatch") }
            guard try loaded.getData(path: "/index.html") == indexHtml else { throw Fail("index.html data mismatch") }
        }

        await testAsync("Read/Write: file roundtrip") {
            let tmpDir = FileManager.default.temporaryDirectory
                .appendingPathComponent("wvb-test-\(Int.random(in: 0..<Int.max))")
            try FileManager.default.createDirectory(at: tmpDir, withIntermediateDirectories: true)
            defer { try? FileManager.default.removeItem(at: tmpDir) }

            let builder = BundleBuilder(version: nil)
            _ = try builder.insertEntry(path: "/index.js", data: indexJs, contentType: nil, headers: nil)
            _ = try builder.insertEntry(path: "/index.html", data: indexHtml, contentType: nil, headers: nil)
            let bundle = try builder.build(options: nil)

            let filePath = tmpDir.appendingPathComponent("bundle.wvb").path
            _ = try await writeBundle(bundle: bundle, filepath: filePath)
            guard FileManager.default.fileExists(atPath: filePath) else { throw Fail("file not created") }

            let loaded = try await readBundle(filepath: filePath)
            guard loaded.descriptor().header().version() == .v1 else { throw Fail("version mismatch") }
            let index = loaded.descriptor().index()
            guard index.entries().count == 2 else { throw Fail("expected 2 entries") }
            guard index.containsPath(path: "/index.js") else { throw Fail("missing /index.js") }
            guard try loaded.getData(path: "/index.js") == indexJs else { throw Fail("index.js data mismatch") }
        }

        // ── Protocol ─────────────────────────────────────────────────────
        await testAsync("BundleProtocolHandler: 200 index.html") {
            try await withSource { source in
                let handler = BundleProtocolHandler(source: source)
                let response = try await handler.handle(method: .get, uri: "https://app.wvb/index.html", headers: nil)
                guard response.status == 200 else { throw Fail("status=\(response.status)") }
                guard !response.body.isEmpty else { throw Fail("body should not be empty") }
                guard response.headers["content-type"] != nil else { throw Fail("content-type header missing") }
            }
        }

        await testAsync("BundleProtocolHandler: 200 root redirect") {
            try await withSource { source in
                let handler = BundleProtocolHandler(source: source)
                let response = try await handler.handle(method: .get, uri: "https://app.wvb/", headers: nil)
                guard response.status == 200 else { throw Fail("status=\(response.status)") }
            }
        }

        await testAsync("BundleProtocolHandler: 404 not found") {
            try await withSource { source in
                let handler = BundleProtocolHandler(source: source)
                let response = try await handler.handle(method: .get, uri: "https://app.wvb/not_found.html", headers: nil)
                guard response.status == 404 else { throw Fail("status=\(response.status)") }
            }
        }

        await testAsync("BundleProtocolHandler: HEAD 200") {
            try await withSource { source in
                let handler = BundleProtocolHandler(source: source)
                let response = try await handler.handle(method: .head, uri: "https://app.wvb/index.html", headers: nil)
                guard response.status == 200 else { throw Fail("status=\(response.status)") }
                guard response.body.isEmpty else { throw Fail("HEAD response body must be empty") }
            }
        }

        // ── Builtin bundle ───────────────────────────────────────────────
        await testAsync("Source: fetch builtin bundle") {
            try await withBuiltinSource { source in
                let bundle = try await source.fetchBundle(bundleName: "app")
                guard try bundle.getData(path: "/index.html") != nil else {
                    throw Fail("index.html not found in builtin bundle")
                }
            }
        }

        await testAsync("Source: fetchDescriptor builtin") {
            try await withBuiltinSource { source in
                let descriptor = try await source.fetchDescriptor(bundleName: "app")
                guard descriptor.containsPath(path: "/index.html") else {
                    throw Fail("index.html not in descriptor")
                }
            }
        }

        await testAsync("BundleProtocolHandler: builtin 200 index.html") {
            try await withBuiltinSource { source in
                let handler = BundleProtocolHandler(source: source)
                let response = try await handler.handle(method: .get, uri: "https://app.wvb/index.html", headers: nil)
                guard response.status == 200 else { throw Fail("status=\(response.status)") }
                guard !response.body.isEmpty else { throw Fail("body should not be empty") }
            }
        }

        await testAsync("BundleProtocolHandler: exact path resolver does not rewrite to index.html") {
            try await withSource { source in
                let options = BundleProtocolOptions(pathResolver: .exact)
                let handler = BundleProtocolHandler(source: source, options: options)
                let served = try await handler.handle(method: .get, uri: "https://app.wvb/index.html", headers: nil)
                guard served.status == 200 else { throw Fail("status=\(served.status)") }
                let response = try await handler.handle(method: .get, uri: "https://app.wvb/", headers: nil)
                guard response.status == 404 else { throw Fail("status=\(response.status)") }
            }
        }

        await testAsync("BundleProtocolHandler: allowWvbSuffixOnly rejects other hosts") {
            try await withSource { source in
                let options = BundleProtocolOptions(
                    bundleResolver: .hostname(segment: .first, allowWvbSuffixOnly: true)
                )
                let handler = BundleProtocolHandler(source: source, options: options)
                let served = try await handler.handle(method: .get, uri: "https://app.wvb/index.html", headers: nil)
                guard served.status == 200 else { throw Fail("status=\(served.status)") }
                var didThrow = false
                do {
                    _ = try await handler.handle(method: .get, uri: "https://app.example.com/index.html", headers: nil)
                } catch {
                    didThrow = true
                }
                guard didThrow else { throw Fail("expected error for a host without the .wvb suffix") }
            }
        }

        test("ProxyProtocolHandler: init") {
            _ = ProxyProtocolHandler(hosts: ["myapp": "http://localhost:9999"])
        }

        await testAsync("BundleProtocolHandler: a request body is accepted") {
            try await withSource { source in
                let handler = BundleProtocolHandler(source: source)
                // The bundle protocol serves GET/HEAD only, but the body still travels over the FFI.
                let response = try await handler.handle(
                    method: .post,
                    uri: "https://app.wvb/index.html",
                    headers: nil,
                    body: Data(#"{"hello":"world"}"#.utf8)
                )
                guard response.status == 405 else { throw Fail("status=\(response.status)") }
            }
        }

        await testAsync("ProxyProtocolHandler: unknown host error") {
            let handler = ProxyProtocolHandler(hosts: ["known": "http://localhost:9999"])
            var didThrow = false
            do {
                _ = try await handler.handle(method: .get, uri: "https://unknown.wvb/index.html", headers: nil)
            } catch {
                didThrow = true
            }
            guard didThrow else { throw Fail("expected error for unknown host") }
        }

        await testAsync("ProxyProtocolHandler: custom resolver receives the uri") {
            let resolver = RecordingProxyResolver()
            let handler = ProxyProtocolHandler.custom(resolver: resolver)
            var didThrow = false
            do {
                _ = try await handler.handle(method: .get, uri: "https://app.wvb/index.html", headers: nil)
            } catch {
                didThrow = true
            }
            guard didThrow else { throw Fail("expected error when the resolver returns nil") }
            guard resolver.seen == "https://app.wvb/index.html" else {
                throw Fail("resolver saw uri=\(resolver.seen ?? "none")")
            }
        }

        results = newResults
        isRunning = false
    }

    private func withBuiltinSource(_ block: (Source) async throws -> Void) async throws {
        print("resource url: \(Bundle.main.resourceURL?.absoluteString ?? "none")")
        guard let builtinDir = Bundle.main.resourceURL?.appendingPathComponent("assets").appendingPathComponent("bundles").appendingPathComponent("builtin") else {
            throw Fail("builtin resource directory not found in app bundle")
        }
        let remoteDir = FileManager.default.temporaryDirectory
            .appendingPathComponent("wvb-builtin-remote-\(Int.random(in: 0..<Int.max))")
        try FileManager.default.createDirectory(at: remoteDir, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: remoteDir) }

        let emptyManifest = #"{"manifestVersion":1,"bundles":{}}"#
        try Data(emptyManifest.utf8).write(to: remoteDir.appendingPathComponent("manifest.json"))

        let source = try Source(config: SourceConfig(
            builtinDir: builtinDir.path,
            remoteDir: remoteDir.path,
            builtinManifestFilepath: nil,
            remoteManifestFilepath: nil
        ))
        try await block(source)
    }

    private func withSource(_ block: (Source) async throws -> Void) async throws {
        let tmpDir = FileManager.default.temporaryDirectory
            .appendingPathComponent("wvb-fixture-\(Int.random(in: 0..<Int.max))")
        let remoteDir = tmpDir.appendingPathComponent("remote")
        let appDir = remoteDir.appendingPathComponent("app")
        try FileManager.default.createDirectory(at: appDir, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: tmpDir) }

        let manifest = #"{"manifestVersion":1,"bundles":{"app":{"versions":{"1.0.0":{}},"currentVersion":"1.0.0"}}}"#
        try Data(manifest.utf8).write(to: remoteDir.appendingPathComponent("manifest.json"))

        let builder = BundleBuilder(version: nil)
        _ = try builder.insertEntry(path: "/index.html", data: indexHtml, contentType: "text/html", headers: nil)
        let bundle = try builder.build(options: nil)
        let bundleBytes = try writeBundleToBytes(bundle: bundle)
        try bundleBytes.write(to: appDir.appendingPathComponent("1.0.0.wvb"))

        let builtinDir = tmpDir.appendingPathComponent("builtin")
        try FileManager.default.createDirectory(at: builtinDir, withIntermediateDirectories: true)

        let source = try Source(config: SourceConfig(
            builtinDir: builtinDir.path,
            remoteDir: remoteDir.path,
            builtinManifestFilepath: nil,
            remoteManifestFilepath: nil
        ))
        try await block(source)
    }
}

private struct Fail: LocalizedError {
    let errorDescription: String?
    init(_ message: String) { errorDescription = message }
}

/// Records the uri it is asked to resolve, and never proxies.
private final class RecordingProxyResolver: ProxyResolver, @unchecked Sendable {
    private let lock = NSLock()
    private var uri: String?

    var seen: String? {
        lock.lock()
        defer { lock.unlock() }
        return uri
    }

    func resolve(uri: String) async -> String? {
        lock.lock()
        self.uri = uri
        lock.unlock()
        return nil
    }
}
